File: searching/linear_search/python/linear_search.py
"""線形探索（Linear Search）の実装。

配列の先頭から末尾まで要素を1つずつ順番に比較していき、targetと一致する
要素が見つかればそのインデックスを返す。配列がソートされている必要はなく、
任意の配列に適用できる。

時間複雑度: 最良 O(1)（先頭要素が一致する場合）、平均/最悪 O(n)
空間複雑度: O(1)
"""
from typing import List


def linear_search(arr: List[int], target: int) -> int:
    """targetが見つかった場合はそのインデックス（0始まり。複数該当する場合は
    いずれかのインデックス）、見つからなければ-1を返す。"""
    for i, value in enumerate(arr):
        if value == target:
            return i
    return -1

File: searching/linear_search/matlab/linear_search.m
function idx = linear_search(arr, target)
% LINEAR_SEARCH 配列を先頭から順に走査してtargetを探す（線形探索）
%
%   配列の先頭から末尾まで要素を1つずつ順番に比較していき、targetと
%   一致する要素が見つかればそのインデックスを返す。配列がソートされている
%   必要はなく、任意の配列に適用できる。
%
%   時間複雑度: 最良 O(1)（先頭要素が一致する場合）、平均/最悪 O(n)
%   空間複雑度: O(1)
%
%   idx = LINEAR_SEARCH(arr, target) は、targetが見つかった場合は
%   1始まりのインデックス（複数該当する場合はいずれかのインデックス）を、
%   見つからなければ-1を返す。

    idx = -1;
    for i = 1:length(arr)
        if arr(i) == target
            idx = i;
            return;
        end
    end
end
